sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java

```diff
@@ -482,40 +506,62 @@ private void captureSessionToken(RxDocumentServiceRequest request, Map<String, S
+        // Master resource operations don't require session token.
+        if (isMasterOperation(request.getResourceType(), request.getOperationType())) {
+            if (!Strings.isNullOrEmpty(request.getHeaders().get(HttpConstants.HttpHeaders.SESSION_TOKEN))) {
                 request.getHeaders().remove(HttpConstants.HttpHeaders.SESSION_TOKEN);
             }
-            return; //User is explicitly controlling the session.
+            return Mono.empty();
         }
 
-        if (!sessionTokenApplicable || isMasterOperation(request.getResourceType(), request.getOperationType())) {
-            return; // Only apply the session token in case of session consistency and when resource is not a master resource
+        if (!Strings.isNullOrEmpty(request.getHeaders().get(HttpConstants.HttpHeaders.SESSION_TOKEN))) {
+            return Mono.empty(); // User is explicitly controlling the session.
         }
 
-        //Apply the ambient session.
-        String sessionToken = this.sessionContainer.resolveGlobalSessionToken(request);
+        boolean sessionConsistency = RequestHelper.getConsistencyLevelToUse(this.gatewayServiceConfigurationReader,
+            request) == ConsistencyLevel.SESSION;
 
-        if (!Strings.isNullOrEmpty(sessionToken)) {
-            headers.put(HttpConstants.HttpHeaders.SESSION_TOKEN, sessionToken);
+        if (!sessionConsistency ||
+            (!request.isReadOnlyRequest() && request.getOperationType() != OperationType.Batch && !this.useMultipleWriteLocations)) {
+            return Mono.empty();
+            // Only apply the session token in case of session consistency and if request is read only,
+            // apply token for write request only if batch operation or multi master
+        }
+
+        String partitionKeyRangeId = request.getHeaders().get(HttpConstants.HttpHeaders.PARTITION_KEY_RANGE_ID);
+        PartitionKeyInternal partitionKeyInternal = request.getPartitionKeyInternal();
+
+        if (StringUtils.isNotEmpty(partitionKeyRangeId)) {
+            SessionTokenHelper.setPartitionLocalSessionToken(request, partitionKeyRangeId, sessionContainer);
+            return Mono.empty();
+        } else if (partitionKeyInternal != null) {
+            return this.collectionCache.resolveCollectionAsync(BridgeInternal.getMetaDataDiagnosticContext(request.requestContext.cosmosDiagnostics), request).
+                flatMap(collectionValueHolder -> partitionKeyRangeCache.tryLookupAsync(BridgeInternal.getMetaDataDiagnosticContext(request.requestContext.cosmosDiagnostics),
+                    collectionValueHolder.v.getResourceId(),
+                    null,
+                    null).flatMap(collectionRoutingMapValueHolder -> {
+                    String effectivePartitionKeyString = PartitionKeyInternalHelper
+                        .getEffectivePartitionKeyString(
+                            partitionKeyInternal,
+                            collectionValueHolder.v.getPartitionKey());
```

Comment:
NPE check, how can we be sure `collectionValueHolder.v` is not null? 

Collection delete scenario, or collection delete, recreate with the same name?

Reply:
Recreate scenario is added in https://github.com/Azure/azure-sdk-for-java/pull/25811.
If cache is null, sdk wont able to serve the any request irrespective of this code. Checked other places, some are not doing null, some doing and throwing illegalException